### ISSUE TITLE
GPU/CPU model synchronization

### ISSUE THREAD

**Issue by E1eMenta** [pytorch/pytorch#6228]
- PyTorch 0.3.1:
- Ubuntu 16.04.3 LTS:
- PyTorch installed via pip
- Anaconda Python 3.5.3:
- CUDA 9.0/cuDNN 7.0 version:
- gtx 1080ti:

I'm developing asynchronous multi-gpu. For this task I forward/backward gpu model, copy gradients to shared model, optimize, and copy weights back. Code diverges on gpu and converges on cpu. What the reason?

```
import torch
import torch.nn as nn
import torch.nn.functional as F
from torchvision import datasets, transforms

class Net(nn.Module):
    def __init__(self):
        super(Net,self).__init__()
        self.conv1 = nn.Conv2d(1, 10, kernel_size=5)
        self.conv2 = nn.Conv2d(10, 20, kernel_size=5)
        self.conv2_drop = nn.Dropout2d()
        self.fc1 = nn.Linear(320, 50)
        self.fc2 = nn.Linear(50, 10)

    def forward(self, x):
        x = F.relu(F.max_pool2d(self.conv1(x), 2))
        x = F.relu(F.max_pool2d(self.conv2_drop(self.conv2(x)), 2))
        x = x.view(-1, 320)
        x = F.relu(self.fc1(x))
        x = F.dropout(x, training=self.training)
        x = self.fc2(x)
        return F.log_softmax(x, dim=1)

train_loader = torch.utils.data.DataLoader(
                datasets.MNIST('data', train=True, download=True,
                               transform=transforms.Compose([
                                   transforms.ToTensor(),
                                   transforms.Normalize((0.1307,), (0.3081,))
                               ])),
                batch_size=512, shuffle=True, num_workers=1)

model = Net().cuda()
shared_model = Net()

optimizer = torch.optim.SGD(shared_model.parameters(), lr=0.01, momentum=0.5)

for param_group in optimizer.param_groups:
    params = param_group['params']

# Grads from None to good value
weights_mean = [torch.mean(param) for param in params]
weights_sum = sum(weights_mean)
weights_sum.backward()

model.train()
epoch = 0
loss_sum = 0

while True:
    for batch_idx, (data, target) in enumerate(train_loader):
        model.load_state_dict(shared_model.state_dict())

        data, target = torch.autograd.Variable(data.cuda()), torch.autograd.Variable(target.cuda())

        optimizer.zero_grad()

        output = model(data)
        loss = F.nll_loss(output, target)
        loss.backward()

        for param, shared_parameter in zip(model.parameters(), shared_model.parameters()):
                shared_parameter.grad.data = param.grad.data.cpu()

        loss_sum += loss
        optimizer.step()

        loss_sum += loss
        optimizer.step()
        if batch_idx % 10 == 0:
            print('Train Epoch: {} Loss: {:.6f}'.format(
                epoch, loss.data[0] / 10))
            loss_sum = 0
    epoch += 1
```
